## Basic Python - Project <a id='intro'></a>

## Introduction <a id='intro'></a>
In this project, we'll be analyzing data from the entertainment industry. We'll research the "Golden Age" of television by studying the records on movies and shows.Our research will focus on the the years 1999 to present day when the "Golden Age" of television started.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. The source file is not available, unfortuantely. There is no information about the quality of the data, so you we will need to explore it before doing the analysis.

First, we'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing,  try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

## Stage 1. Data overview <a id='data_review'></a>

In [1]:
# importing pandas
import pandas as pd

Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [2]:
# Loading the data
df = pd.read_csv('/datasets/movies_and_shows.csv')

In [3]:
# Viewing first 10 rows of the data
df.head(10)

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

In [4]:
# Obtaining general information about the data in df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>

In [5]:
# Viewing list of column names in the df table
df.columns

Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

In [6]:
# We'll change the column names according to standard format:
df = df.rename(columns={
    '   name':'name', 'Character':'character', 'r0le':'role', 'TITLE':'title',   '  Type':'type', 'release Year': 'release_year', 'imdb sc0re':'imdb_score', 'imdb v0tes':'imdb_votes'})# renaming columns


In [7]:
# Confirming changes
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

In [8]:
# Checming missing values
df.isna().sum()

name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` *and* `'imdb_votes'` columns.

In [9]:
# Dropping rows with missing values
df = df.dropna(subset = ['title', 'imdb_score', 'imdb_votes']) # dropping rows where columns with title, scores and votes have missing values


In [10]:
# Confirming changes
df.isna().sum()


name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

In [11]:
# Chescking for duplicates
df.duplicated().sum()

6994

Review the duplicate rows to determine if removing them would distort our dataset.

In [12]:
# Checking duplicate rows
df[df.duplicated()]

name                  character      role  \
7561         Philip Greene  Baseball Fan (uncredited)     ACTOR   
14512             Dan Levy                   Reporter     ACTOR   
18952     Nicolas Le Nev??                    unknown  DIRECTOR   
22456    John Iii Franklin                    Himself     ACTOR   
29557         Claudio Roca                   Nicol?­s     ACTOR   
...                    ...                        ...       ...   
85569       Jessica Cediel            Liliana Navarro     ACTOR   
85570  Javier Gardeaz?­bal   Agust??n "Peluca" Ort??z     ACTOR   
85571        Carla Giraldo             Valery Reinoso     ACTOR   
85572  Ana Mar??a S?­nchez                    Lourdes     ACTOR   
85577         Isabel Gaona                     Cacica     ACTOR   

                                 title   type  release_year  \
7561                   How Do You Know  MOVIE          2010   
14512  A Very Harold & Kumar Christmas  MOVIE          2011   
18952                       Sammy & Co   SHOW          2014   
22456                    Last Chance U   SHOW          2016   
29557                   Narcos: Mexico   SHOW          2018   
...                                ...    ...           ...   
85569                          Lokillo  MOVIE          2021   
85570                          Lokillo  MOVIE          2021   
85571                          Lokillo  MOVIE          2021   
85572                          Lokillo  MOVIE          2021   
85577                          Lokillo  MOVIE          2021   

                                 genres  imdb_score  imdb_votes  
7561     ['comedy', 'drama', 'romance']         5.4     50383.0  
14512  ['comedy', 'fantasy', 'romance']         6.2     69562.0  
18952         ['animation', 'european']         5.7        31.0  
22456        ['documentation', 'sport']         8.4      6897.0  
29557                ['drama', 'crime']         8.4     82042.0  
...                                 ...         ...         ...  
85569                        ['comedy']         3.8        68.0  
85570                        ['comedy']         3.8        68.0  
85571                        ['comedy']         3.8        68.0  
85572                        ['comedy']         3.8        68.0  
85577                        ['comedy']         3.8        68.0  

[6994 rows x 9 columns]

The duplicate data will distort our results. We'll remove them from the dataframe.

In [13]:
# Removing duplicates
df = df.drop_duplicates().reset_index(drop=True)

In [14]:
# Confirming duplicates removal
df.duplicated().sum()

0

In [15]:
# Checking for implicit `type` value duplicates
df['type'].sort_values().unique()

array(['MOVIE', 'SHOW', 'movies', 'shows', 'the movie', 'tv', 'tv series',
       'tv show', 'tv shows'], dtype=object)

We'll address these implicit duplicates by defining the `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function will correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [16]:
def replace_wrong_show(df, wrong_shows_list, correct_show):
    df['type'] = df['type'].replace(wrong_shows_list, correct_show)
    return df

In [17]:
# Storing wrong show `tyoe` values and assigning to variable for function
wrong_shows_list = ['tv', 'tv series', 'tv shows', 'tv show', 'shows']

# Declaring correct value for function
correct_show = 'SHOW'

# Replacing the values by calling the function with the appropriate parameters
replace_wrong_show(df, wrong_shows_list, correct_show)

name                character      role        title  \
0          Robert De Niro            Travis Bickle     ACTOR  Taxi Driver   
1            Jodie Foster            Iris Steensma     ACTOR  Taxi Driver   
2           Albert Brooks                      Tom     ACTOR  Taxi Driver   
3           Harvey Keitel  Matthew 'Sport' Higgins     ACTOR  Taxi Driver   
4         Cybill Shepherd                    Betsy     ACTOR  Taxi Driver   
...                   ...                      ...       ...          ...   
73854       A??da Morales                  Maritza     ACTOR      Lokillo   
73855    Adelaida Buscato               Mar??a Paz     ACTOR      Lokillo   
73856  Luz Stella Luengas             Karen Bayona     ACTOR      Lokillo   
73857        In??s Prieto                    Fanny     ACTOR      Lokillo   
73858      Julian Gaviria                  unknown  DIRECTOR      Lokillo   

            type  release_year              genres  imdb_score  imdb_votes  
0          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
1          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
2          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
3          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
4          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
...          ...           ...                 ...         ...         ...  
73854  the movie          2021          ['comedy']         3.8        68.0  
73855  the movie          2021          ['comedy']         3.8        68.0  
73856  the movie          2021          ['comedy']         3.8        68.0  
73857  the movie          2021          ['comedy']         3.8        68.0  
73858  the movie          2021          ['comedy']         3.8        68.0  

[73859 rows x 9 columns]

In [18]:
# Confirming changes
print(df['type'].unique())

['MOVIE' 'the movie' 'SHOW' 'movies']


### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, we can now define how the assumption will be checked. We'll calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

In [19]:
# Filtering the data to only include films from 1999 to present
df_shows = df[df['release_year'] >= 1999] 

In [20]:
# Repeating conditional indexing to extract rows with 'SHOW' values in the `type` column
df_shows = df_shows[df_shows['type'] == 'SHOW']

In [21]:
# Rounding the `imdb_score` values
df_shows['imdb_score'] = df_shows['imdb_score'].round()

# Confirming changes by viewing the last 5 rows
df_shows.tail()

name character      role      title  type  release_year  \
73831     Maneerat Kam-Uan        Ae     ACTOR  Let's Eat  SHOW          2021   
73832    Rudklao Amratisha   unknown     ACTOR  Let's Eat  SHOW          2021   
73833      Jaturong Mokjok   unknown     ACTOR  Let's Eat  SHOW          2021   
73834     Pisamai Wilaisak   unknown     ACTOR  Let's Eat  SHOW          2021   
73835  Sarawut Wichiensarn   unknown  DIRECTOR  Let's Eat  SHOW          2021   

                    genres  imdb_score  imdb_votes  
73831  ['drama', 'comedy']         8.0         5.0  
73832  ['drama', 'comedy']         8.0         5.0  
73833  ['drama', 'comedy']         8.0         5.0  
73834  ['drama', 'comedy']         8.0         5.0  
73835  ['drama', 'comedy']         8.0         5.0

In [22]:
# Grouping the data by scores to get a count of each unique score value to identify outliers
df_shows.groupby('imdb_score')['imdb_votes'].count()

imdb_score
2.0       24
3.0       27
4.0      180
5.0      592
6.0     2494
7.0     4706
8.0     4842
9.0      557
10.0       8
Name: imdb_votes, dtype: int64

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

We'll use conditional filtering to remove the outliers.

In [23]:
# Filtering out rows with `imdb_score` values less than 4
df_shows = df_shows[df_shows['imdb_score'] >= 4]# filter dataframe using two conditions (scores to be in the range 4-9)

# Filtering out rows with `imbd_score` values greater than 9
df_shows = df_shows[df_shows['imdb_score'] <=9]

# Grouping the data by unqique `imdb_score` values to get the average number of votes for each unique score
df_shows = df_shows.groupby('imdb_score')['imdb_votes'].mean().reset_index()

In [24]:
# Ronding the average imdb votes values
df_shows['imdb_votes'] = df_shows['imdb_votes'].round(0)

In [25]:
# Renaming columns
df_shows = df_shows.rename(columns={'imdb_votes':'average_votes'})

In [26]:
# Printing the scores in descending order by average number of votes
df_shows.sort_values(by='average_votes', ascending=False)

imdb_score  average_votes
5         9.0       126904.0
4         8.0        30299.0
3         7.0         8727.0
0         4.0         5278.0
2         6.0         3482.0
1         5.0         3144.0

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.